src/clic.adoc

```diff
@@ -2053,6 +2053,199 @@ the setting in the CLIC Configuration register (`mcliccfg.xnlbits`).
+
+#TODO: Discuss if we want to have the illegal instruction exceptions as defined here
+
+With these requirements, implementations are allowed to speculatively start executing an ipush instruction in case of a taken interrupt.
+
+#TODO: discuss if we want to enforce this for all interrupts or only HW-vectored interrupts (i.e. coming from xtvt). My assumption is we do not want to do this exceptions
+
+=== Late Preemption of the interrupt handler
+
+While executing the ipush instruction, a hart may optionally check for higher priority/level pending and enabled interrupts at the current privilege mode, and switch to fetching this interrupt handler instead.
+Once the instruction retires, or the interrupts are disabled, no late preemption can occur.
+
+=== Tail chaining of interrupts handlers
+
+While executing the ipopxret instruction, a hart may optionally check for any pending and enabled interrupts at the current privilege mode.
+If such an interrupt exists, the hart may abort executing the ipopxret instruction and trigger the following actions:
+  - Update the xcause.exccode to the corresponding interrupt id.
+  - Perform the trap handler fetch.
+  - Start excuting the trap handler, while skipping the portions of ipush that have not been reverted by the aborted ipopxret.
+
+Explicitly, xepc, xpil, xpp, and xpexccode remain unchanged.
+
+NOTE: In the typical case of tail chaining, the check for another pending enabled interrupt is done after disabling interrupts at the start of the ipopxret instruction. In this case, only the conditional interrupt enablement of the ipush instruction would need to executed.
+
+Once the instruction retires, or the interrupts are disabled, no late preemption can occur.
+
+=== Stack layout
+
+To ease debugging, a stack layout is specified for the saved context.
+
+#TODO: find the best way to specify the memory layout in asciidoc
+
+* xtvec.mode != CLIC:
+** XXLEN = 32:
+
+stack pointer adjustment: 32
+
+24: x15
+
+20: x14
+
+16: x13
+
+12: x12
+
+8: x11
+
+4: x10
+
+** XXLEN = 64:
+
+stack pointer adjustment: 48
+
+48: x15
+
+40: x14
+
+32: x13
+
+24: x12
+
+16: x11
+
+8: x10
+
+* xtvec.mode = CLIC:
+
+start of stackframe remains same. In addition:
+
+** XXLEN = 32
+
+stack pointer adjustment: 32
+
+32: reserved (5) | xpp (2) |  xpie (1) |  xpil (8) | reserved (4) | xpexccode (12)
+
+28: xpepc (XXLEN)
+
+** XXLEN = 32
+
+stack pointer adjustment: 64
+
+60: reserved (5) | xpp (2) |  xpie (1) |  xpil (8) | reserved (4) | xpexccode (12)
+
+56: xpepc (XXLEN)
+
+== Fault handling
+
+According the the Machine Level ISA, trap handlers must be designed to not cause exceptions during their critical section.
+This applies to the ipush and ipopxret instruction.
+Therefore, by design, it must be guaranteed that no faults occur during the execution of these instructions.
+
+=== Debug
+
+#TODO: Define what happens if there is a break point at the ipush instruction address
```

Comment:
If possible, I think we should define this in a way that doesn't prevent speculative state saving before the handler address has been fetched (since we need the table lookup to know if it's a break point hit) as that could add many precious cycles to the interrupt latency.

Reply:
agree. ipush cannot be burdened. 
i do not think there is a good solution to it. i think it will end up in a way that a breakpoint on the ipush instruction may be able to observe a partial update or even complete update of ipush. 
another route would be to not allow breakpoints on ipush, or to say a matching breakpoint will break on the subsequent instruction.

Reply:
Can we add that restriction to breakpoints? In that case it sounds like a reasonable solution. I could see some pushback, but it's worth trying

Reply:
something is in now and ready to be challenged/improved